### ISSUE TITLE
`generate` do not stop after generating `eos_token` with batch process

### ISSUE THREAD

**Issue by notoookay** [huggingface/transformers#31261]
### System Info

- `transformers` version: 4.36.2
- Platform: Linux-3.10.0-1160.el7.x86_64-x86_64-with-glibc2.31
- Python version: 3.10.14
- Huggingface_hub version: 0.22.2
- Safetensors version: 0.4.2
- Accelerate version: 0.30.1
- Accelerate config:    not found
- PyTorch version (GPU?): 2.1.2+cu118 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: no

### Who can help?

@ArthurZucker @gante 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```
prompts = example["prompt"] # (batch_size,)
inputs = tokenizer(prompts,
                  return_tensors="pt",
                  padding=True,
                  truncation=True).to(model.device)
input_ids = inputs["input_ids"]
outputs = model.generate(input_ids,
                             max_new_tokens=max_new_tokens,
                             do_sample=False,
                             eos_token_id=tokenizer.eos_token_id,)
```

### Expected behavior

I found when generating with only one input, model (I use llama2 to test) acts normal, when generating with a batch of inputs, model still generating `<unk>` after `eos_token` has been generated, like this:

`### Instruction:\nAnswer the following question.\n\n### Input:\nWhat is the capital of France?\n\n### Response:\nThe capital of France is Paris.</s><unk><unk><unk><unk><unk>`

Is this the reason: **The `generate` function is designed to generate sequences of a consistent length for all sequences in a batch. When generating sequences in a batch, the `generate` function continues until the longest sequence in the batch reaches the specified `max_length`, or until all sequences have generated an EOS token, whichever comes first. This means that if one sequence in the batch generates an EOS token before the others, the `generate` function will continue generating tokens for that sequence (likely just padding tokens) until the other sequences have finished.**

If it is, will it be fixed in the future? If not, please give me some suggestion.

Thank you.

**Comment by ArthurZucker** [comment]
Hey! This is expected I think. Unless it's actually generating them, I suppose it's just the padding we add to make the batch fit

**Comment by notoookay** [comment]
Got it, thank you.
